### PR TITLE
Add a reference example for PSA_ALG_SP800_108_COUNTER_HMAC/CMAC

### PR DESCRIPTION
**Pull request by mswarowsky**
Adds a reference implementation for a counter mode KDF using the construction recommended by NIST SP 800-108, introduced in #123.

The example is using the PSA API's with mbedtls implementation.

### CODE REVIEW

File: examples/crypto/SP800-108_counter_KDF/main.c [hunk 1]
```diff
@@ -0,0 +1,400 @@
+/*
+ * TODO: Change to the right licence
+ * Copyright (c) 2023 Nordic Semiconductor ASA
```
Comment (athoelke) [inline examples/crypto/SP800-108_counter_KDF/main.c:3]: Please use an SPDX file copyright tag, either to replace or in addition to this comment. E.g.

```suggestion
 * SPDX-FileCopyrightText: Copyright (c) 2023 Nordic Semiconductor ASA
```

File: examples/crypto/SP800-108_counter_KDF/README.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
+NIST SP 800-108 counter-mode KDF
```
[the remark this answers is not in the document]
Comment (mswarowsky) [reply examples/crypto/SP800-108_counter_KDF/README.md:1]: Ok added it

File: examples/crypto/SP800-108_counter_KDF/main.c [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+	if (key_buffer_length != PSA_HASH_LENGTH(alg)) {
+		printf("\nWrong key size, expected: %u, got:%lu\n", PSA_HASH_LENGTH(alg),
+		       key_buffer_length);
+		return PSA_ERROR_INVALID_ARGUMENT;
+	}
+
+	/* Configure the input key attributes */
+	psa_set_key_usage_flags(&key_attributes,
+				PSA_KEY_USAGE_VERIFY_HASH | PSA_KEY_USAGE_SIGN_HASH);
+	psa_set_key_lifetime(&key_attributes, PSA_KEY_LIFETIME_VOLATILE);
```
[the remark this answers is not in the document]
Comment (mswarowsky) [reply examples/crypto/SP800-108_counter_KDF/main.c]: Removed it

File: examples/crypto/SP800-108_counter_KDF/main.c [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+
+	if (key_buffer_length != PSA_HASH_LENGTH(alg)) {
+		printf("\nWrong key size, expected: %u, got:%lu\n", PSA_HASH_LENGTH(alg),
+		       key_buffer_length);
+		return PSA_ERROR_INVALID_ARGUMENT;
+	}
+
+	/* Configure the input key attributes */
+	psa_set_key_usage_flags(&key_attributes,
+				PSA_KEY_USAGE_VERIFY_HASH | PSA_KEY_USAGE_SIGN_HASH);
```
[the remark this answers is not in the document]
Comment (mswarowsky) [reply examples/crypto/SP800-108_counter_KDF/main.c]: changed it

File: examples/crypto/SP800-108_counter_KDF/main.c [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+	if (status != PSA_SUCCESS) {
+		printf("\npsa_mac_sign_setup failed! (Error: %d)\n", status);
+		return status;
+	}
+
+	status =
+		mac_update_fixed_input(&operation, label, label_length, context, context_length, L);
+	if (status != PSA_SUCCESS) {
+		printf("\nmac_update_fixed_input failed! (Error: %d)\n", status);
+		return status;
```
[the remark this answers is not in the document]
Comment (mswarowsky) [reply examples/crypto/SP800-108_counter_KDF/main.c]: I added an error block at the end, that calls `psa_mac_abort` the operation

File: examples/crypto/SP800-108_counter_KDF/main.c [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+		}
+
+		/* counter is needed as 4 byte big-endian */
+		uint32_t counter_be = le32_to_be32(i);
+
+		/* [i]_4  */
+		status = psa_mac_update(&operation, (uint8_t *)&counter_be, sizeof(counter_be));
+		if (status != PSA_SUCCESS) {
+			printf("\npsa_mac_update L failed! (Error: %d)\n", status);
+			return status;
```
[the remark this answers is not in the document]
Comment (mswarowsky) [reply examples/crypto/SP800-108_counter_KDF/main.c]: Also added a goto error block

File: examples/crypto/SP800-108_counter_KDF/main.c [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
```
[the remark this answers is not in the document]
Comment (mswarowsky) [reply examples/crypto/SP800-108_counter_KDF/main.c]: Also added error block here

**Comment by athoelke** [comment]
If we are going to be concerned with ensuring that the error clean-up code is correct in this example, we should probably also ensure that the example destroys the temporary keys that are created on each call to `sp800_108_counter_hmac_kdf()` and `sp800_108_counter_cmac_kdf()` - on both the success and error paths?

File: examples/crypto/SP800-108_counter_KDF/README.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        cmake --build .
+
+
+Then run the sample
+
+        ./kbkdf
+        Deriving a key using SP800-108 HMAC counter mode...Done
+        Key:
+        ----------------------------------- len: 16 -----------------------------------
+        0x01 0x02 0x03 0x04 0x05 0x06 0x07 0x08 0x09 0x0a 0x0b 0x0c 0x0d 0x0e 0x0d 0x0f
```
Comment (bheeb) [inline examples/crypto/SP800-108_counter_KDF/README.md:38]: This is not correct. The 32 byte HMAC keys from main.c must be used instead.

Comment (mswarowsky) [reply examples/crypto/SP800-108_counter_KDF/README.md:39]: Oh yeah, thanks, die print out was not updated, fixed it

**Comment by mswarowsky** [comment]
Rebased again, is there anything more I can do to get this moving forward?

**Comment by athoelke** [comment]
> Rebased again, is there anything more I can do to get this moving forward?

Hi @mswarowsky - as the example is based on using the mbedtls implementation, I was hoping to have one of the mbedtls team review the example as well.

Let me check if that is likely to happen soon - if not, then I won't let that block merging the PR.

**Comment by gilles-peskine-arm** [comment]
This is still on my list, I'll try to review this week.

File: examples/crypto/SP800-108_counter_KDF/README.md [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        ----------------------------------- len: 32 -----------------------------------
+        0x50 0x53 0x41 0x5f 0x41 0x4c 0x47 0x5f 0x53 0x50 0x38 0x30 0x30 0x5f 0x31 0x30
+        0x38 0x5f 0x43 0x4f 0x55 0x4e 0x54 0x45 0x52 0x20 0x53 0x61 0x6d 0x70 0x6c 0x65
+        ------------------------------------- end -------------------------------------
+        Capacity: 0x2a
+        CMAC without context derived key:
+        ----------------------------------- len: 42 -----------------------------------
+        0xe1 0xec 0xfc 0000 0x1e 0x2e 0x9a 0xdb 0xd0 0x16 0xb3 0xb4 0xf3 0x23 0xce 0000
+        0xc1 0x05 0x82 0xec 0x81 0xe1 0xfc 0x19 0x40 0x47 0x4c 0xa6 0x84 0xf9 0xe5 0x07
+        0xb5 0x8a 0xbd 0x03 0xbc 0xe5 0x23 0x82 0x05 0x11
```
Comment (gilles-peskine-arm) [inline examples/crypto/SP800-108_counter_KDF/README.md:123]: Info: I wrote a [Python implementation](https://github.com/Mbed-TLS/mbedtls/pull/8698) and got the same data.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I've reviewed the code and the documentation. I won't review the build script because I don't know CMake.

I reviewed to the standards of Mbed TLS except for the code style. We may not have the same goals in the PSA example repository, so I don't necessarily expect all my comments to be addressed. I would prefer to at least resolve the comments marked [MAJOR], but Andrew has the final say.

**Review by mswarowsky — COMMENTED** [review]
Addressed the feedback from @gilles-peskine-arm

File: examples/crypto/SP800-108_counter_KDF/main.c [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+
+#include <string.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <psa/crypto.h>
+
+#define CMAC_KEY_LENGTH  (16)
+#define OTUPUT_LENGTH    (42)
```
Comment (athoelke) [inline examples/crypto/SP800-108_counter_KDF/main.c:15]: Spelling of OUTPUT:

```suggestion
#define OUTPUT_LENGTH    (42)
```

File: examples/crypto/SP800-108_counter_KDF/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,47 @@
+#
+# SPDX-FileCopyrightText: Copyright (c) 2023 Nordic Semiconductor ASA
```
Comment (athoelke) [inline examples/crypto/SP800-108_counter_KDF/CMakeLists.txt:2]: The copyright date now includes 2024:

```suggestion
# SPDX-FileCopyrightText: Copyright (c) 2023-2024 Nordic Semiconductor ASA
```

File: examples/crypto/SP800-108_counter_KDF/main.c [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+/* Buffer to hold the key generated from CMAC_CTR_KDF */
+static uint8_t m_output[OTUPUT_LENGTH] = {0};
+
+static size_t min(size_t a, size_t b){
+	return a < b ? a : b;
+}
+
+static void print_hex(const uint8_t *out, const size_t out_length)
+{
+	printf("----------------------------------- len: %lu -----------------------------------\n",
```
Comment (athoelke) [inline examples/crypto/SP800-108_counter_KDF/main.c:49]: This is now 32 characters longer than the other output lines, after the reformatting of the hex bytes output.

File: examples/crypto/SP800-108_counter_KDF/main.c [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+	       out_length);
+	printf("%02x ", out[0]);
+	for (size_t i = 1; i < out_length; i++) {
+		if (i % 16 == 0) {
+			printf("\n");
+		}
+		printf("%02x ", out[i]);
+	}
+	printf("\n------------------------------------- end "
```
Comment (athoelke) [inline examples/crypto/SP800-108_counter_KDF/main.c:58]: This has the same issue as line 49

**Review by athoelke — COMMENTED** [review]
A couple of minor issues. The rework looks correct, Gilles might like to review the update.

As this PR introduces new files, squashing the PR makes incremental review more challenging. Please can you avoid squashing until the review is complete. I will rebase and squash the PR immediately prior to merging, as we don't need/want the edit history in these files..

File: examples/crypto/SP800-108_counter_KDF/main.c [hunk 14]
```diff
@@ -0,0 +1,397 @@
...
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <psa/crypto.h>
+
+#define CMAC_KEY_LENGTH  (16)
+#define OTUPUT_LENGTH    (42)
+#define AES_BLOCK_SIZE    PSA_BLOCK_CIPHER_BLOCK_LENGTH(PSA_KEY_TYPE_AES)
+#define L_LENGTH 	(sizeof(uint32_t))
```
Comment (gilles-peskine-arm) [inline examples/crypto/SP800-108_counter_KDF/main.c:17]: It's defined as 4 in the specification. The fact that it's `sizeof(uint32_t)` is a coincidence.

File: examples/crypto/SP800-108_counter_KDF/main.c [hunk 15]
```diff
@@ -0,0 +1,397 @@
...
+						  0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0d, 0x0f};
+
+static uint8_t m_input_hmac_key[] = {
+	0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b,
+	0x0c, 0x0d, 0x0e, 0x0d, 0x0f, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06,
+	0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0d, 0x0f};
+
+/* Label that identifies the purpose for the derived keying material */
+static uint8_t m_label[] = {"PSA_ALG_SP800_108_COUNTER Sample"};
+/* Don't include the NULL terminator in the length */
```
Comment (gilles-peskine-arm) [inline examples/crypto/SP800-108_counter_KDF/main.c:32]: Minor: here we have a “null terminator” (in a “null-terminated string”), where the word “null” is an ordinary English word (an adjective, meaning “having the value zero”). So “null” should be spelled in lowercase. “`NULL`” is the usual way of writing a _null pointer_ in C code.

**Review by gilles-peskine-arm — APPROVED** [review]
Thanks for addressing my comments. Looks good to me apart from some very minor readability issues.

**Review by mswarowsky — COMMENTED** [review]
Addressed the new comments, with a fixup commit

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Pull request merged**